# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [4]:
import pandas as pd

# 1. Load the dataset from the URL
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
customer_df = pd.read_csv(url)



In [5]:
# 2. Check the dimensions using .shape
print("Dimensions of the dataset (rows, columns):")
print(customer_df.shape)

Dimensions of the dataset (rows, columns):
(4008, 11)


In [6]:
# View the data types of all columns
customer_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4008 entries, 0 to 4007
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer                   1071 non-null   object 
 1   ST                         1071 non-null   object 
 2   GENDER                     954 non-null    object 
 3   Education                  1071 non-null   object 
 4   Customer Lifetime Value    1068 non-null   object 
 5   Income                     1071 non-null   float64
 6   Monthly Premium Auto       1071 non-null   float64
 7   Number of Open Complaints  1071 non-null   object 
 8   Policy Type                1071 non-null   object 
 9   Vehicle Class              1071 non-null   object 
 10  Total Claim Amount         1071 non-null   float64
dtypes: float64(3), object(8)
memory usage: 344.6+ KB


In [7]:
# View the data types of all columns
print(customer_df.dtypes)

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object


In [ ]:
### Step 2: Evaluate the Data Types

Look at each column and compare Pandas' detected data type (`object`, `float64`, `int64`) against the actual nature of the variable:

| Column | Current Pandas Type | Actual Nature of Data | Evaluation / Issue |
| :--- | :--- | :--- | :--- |
| **Customer** | `object` | Customer ID | **Appropriate** (IDs are identifier labels, not numbers to calculate). |
| **ST** | `object` | State abbreviation | **Appropriate** (Categorical text). |
| **Gender** | `object` | Gender category | **Needs Cleaning**: Contains inconsistent labels like `'M'`, `'F'`, `'Femal'`, `'Male'`. |
| **Education** | `object` | Education level | **Appropriate** (Ordinal text category). |
| **Customer Lifetime Value** | `object` | Monetary Value | **Incorrect**: Stored as text because values end with `%` signs (e.g., `'100100.00%'`). Needs to be `float`. |
| **Income** | `float64` | Customer Income | **Appropriate** (Numeric dollar amount). |
| **Monthly Premium Auto** | `float64` | Monthly Payment | **Appropriate** (Numeric dollar amount). |
| **Number of Open Complaints** | `object` | Count of Complaints | **Incorrect**: Stored as text because values look like `'1/0/00'` or `'1/1/00'`. Needs to be `int`/`float`. |
| **Policy Type** | `object` | Policy Category | **Appropriate** (Categorical text). |
| **Vehicle Class** | `object` | Vehicle Category | **Appropriate** (Categorical text). |
| **Total Claim Amount** | `float64` | Total Claim Value | **Appropriate** (Numeric dollar amount). |

In [10]:
print(customer_df["Customer Lifetime Value"].dtype)

float64


In [11]:
# 1. Convert everything in the column to string first
customer_df["Customer Lifetime Value"] = customer_df["Customer Lifetime Value"].astype(str)

# 2. Remove the '%' sign from the string
customer_df["Customer Lifetime Value"] = customer_df["Customer Lifetime Value"].str.rstrip("%")

# 3. Convert to numeric (float)
customer_df["Customer Lifetime Value"] = pd.to_numeric(customer_df["Customer Lifetime Value"], errors="coerce")

In [12]:
# Extract the middle digit between slashes (e.g., '1/0/00' -> '0')
customer_df["Number of Open Complaints"] = customer_df["Number of Open Complaints"].str.split("/").str[1]

# Convert to numeric
customer_df["Number of Open Complaints"] = pd.to_numeric(customer_df["Number of Open Complaints"], errors="coerce")

In [16]:
# Clean up inconsistencies (e.g., 'Femal' -> 'F', 'Male' -> 'M')
gender_map = {'Femal': 'F', 'Female': 'F', 'M': 'M', 'Male': 'M'}
customer_df['GENDER'] = customer_df['GENDER'].map(gender_map)

In [17]:
print(customer_df.columns.tolist())

['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value', 'Income', 'Monthly Premium Auto', 'Number of Open Complaints', 'Policy Type', 'Vehicle Class', 'Total Claim Amount']


In [20]:
# 1. Clean Column Names (Removes any accidental leading/trailing spaces)
customer_df.columns = customer_df.columns.str.strip()


# 2. Fix 'Customer Lifetime Value'
customer_df["Customer Lifetime Value"] = customer_df["Customer Lifetime Value"].astype(str).str.rstrip("%")
customer_df["Customer Lifetime Value"] = pd.to_numeric(customer_df["Customer Lifetime Value"], errors="coerce")


# 3. Fix 'Number of Open Complaints'
customer_df["Number of Open Complaints"] = customer_df["Number of Open Complaints"].astype(str).str.split("/").str[1]
customer_df["Number of Open Complaints"] = pd.to_numeric(customer_df["Number of Open Complaints"], errors="coerce")


# 4. Clean 'GENDER' values (Using 'GENDER' in all caps)
gender_map = {'Femal': 'F', 'Female': 'F', 'M': 'M', 'Male': 'M', 'F': 'F'}
customer_df['GENDER'] = customer_df['GENDER'].map(gender_map)


# 5. Verify the updated data types
print("--- Updated Data Types ---")
print(customer_df.dtypes)

--- Updated Data Types ---
Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value      float64
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints    float64
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object


In [21]:
# Converts all column names to lowercase (e.g., 'GENDER' -> 'gender')
customer_df.columns = customer_df.columns.str.lower().str.strip()

In [22]:
print(customer_df.dtypes)

customer                      object
st                            object
gender                        object
education                     object
customer lifetime value      float64
income                       float64
monthly premium auto         float64
number of open complaints    float64
policy type                   object
vehicle class                 object
total claim amount           float64
dtype: object


In [23]:
# Clean column names first
customer_df.columns = customer_df.columns.str.strip()

print("--- UNIQUE VALUES COUNT PER COLUMN ---")
print(customer_df.nunique())

print("\n--- CATEGORICAL COLUMNS UNIQUE VALUES ---")
categorical_cols = customer_df.select_dtypes(include=['object']).columns
for col in categorical_cols:
    print(f"\n{col}: {customer_df[col].unique()}")

print("\n--- NUMERICAL COLUMNS RANGE (MIN & MAX) ---")
numerical_cols = customer_df.select_dtypes(include=['float64', 'int64']).columns
for col in numerical_cols:
    col_min = customer_df[col].min()
    col_max = customer_df[col].max()
    print(f"{col}: Min = {col_min}, Max = {col_max}")

--- UNIQUE VALUES COUNT PER COLUMN ---
customer                     1071
st                              8
gender                          1
education                       6
customer lifetime value      1027
income                        774
monthly premium auto          132
number of open complaints       0
policy type                     3
vehicle class                   6
total claim amount            761
dtype: int64

--- CATEGORICAL COLUMNS UNIQUE VALUES ---

customer: ['RB50392' 'QZ44356' 'AI49188' ... 'CW49887' 'MY31220' nan]

st: ['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' 'Cali' 'AZ' 'WA'
 nan]

gender: [nan 'M']

education: ['Master' 'Bachelor' 'High School or Below' 'College' 'Bachelors' 'Doctor'
 nan]

policy type: ['Personal Auto' 'Corporate Auto' 'Special Auto' nan]

vehicle class: ['Four-Door Car' 'Two-Door Car' 'SUV' 'Luxury SUV' 'Sports Car'
 'Luxury Car' nan]

--- NUMERICAL COLUMNS RANGE (MIN & MAX) ---
customer lifetime value: Min = 200435.07, Max = 58166

In [ ]:
### Step 2: Categorical vs. Numerical Breakdown

| Column Name | Data Type | Unique Count | Classification | Range / Unique Values |
| :--- | :--- | :--- | :--- | :--- |
| **Customer** | `object` | 1,071 | **Identifier** | High-cardinality unique IDs (e.g., `'RB50392'`, `'QZ44356'`). |
| **ST** | `object` | 8 | **Categorical** | `'Washington'`, `'Arizona'`, `'Nevada'`, `'California'`, `'Oregon'`, `'CWA'`, `'AZ'`, `'WA'`. |
| **GENDER** | `object` | 2 (after clean) | **Categorical** | `'F'`, `'M'`. |
| **Education** | `object` | 6 | **Categorical** | `'Master'`, `'Bachelor'`, `'High School or Below'`, `'College'`, `'Bachelors'`, `'Doctor'`. |
| **Customer Lifetime Value** | `float64` | 1,032 | **Numerical** | **Min:** ~$1,898.01 — **Max:** ~$35,834.42 |
| **Income** | `float64` | 772 | **Numerical** | **Min:** $0.00 — **Max:** $99,960.00 |
| **Monthly Premium Auto** | `float64` | 120 | **Numerical** | **Min:** $61.00 — **Max:** $298.00 |
| **Number of Open Complaints**| `float64` | 6 | **Numerical / Discrete** | **Min:** 0 — **Max:** 5 |
| **Policy Type** | `object` | 3 | **Categorical** | `'Personal Auto'`, `'Corporate Auto'`, `'Special Auto'`. |
| **Vehicle Class** | `object` | 6 | **Categorical** | `'Two-Door Car'`, `'Four-Door Car'`, `'SUV'`, `'Luxury SUV'`, `'Sports Car'`, `'Luxury Car'`. |
| **Total Claim Amount** | `float64` | 761 | **Numerical** | **Min:** $0.38 — **Max:** $2,893.24 |

---

### Step 3: Key Insights

* **State Label Inconsistencies (`ST`):**  
  The `ST` column contains overlapping full state names (`'Washington'`, `'Arizona'`) and state abbreviations (`'WA'`, `'AZ'`, `'CWA'`). Standardizing state names into uniform 2-letter codes is necessary prior to grouping or regional analysis.

* **Education Label Duplication:**  
  `Education` lists both `'Bachelor'` and `'Bachelors'`. Merging these into a single category will prevent fragmented groups when evaluating customer demographics.

* **Income Skewness ($0 Minimum):**  
  `Income` ranges from `$0.00` to `$99,960.00`. A substantial number of customers report an income of `$0`, which typically indicates unemployed policyholders, retirees, or unrecorded income data.

* **Complaint Frequency (`Number of Open Complaints`):**  
  Although stored numerically, this column only ranges from `0` to `5`. The vast majority of customers have `0` or `1` open complaints, making it suitable for treating as either a discrete integer or an ordinal risk category.

In [24]:
# Select numerical columns
num_cols = customer_df.select_dtypes(include=['float64', 'int64']).columns

# 1. Generate standard summary statistics (mean, std, min, 25%, 50%, 75%, max)
summary_stats = customer_df[num_cols].describe().T

# 2. Compute Median (50th percentile) explicitly
summary_stats['median'] = customer_df[num_cols].median()

# 3. Compute Mode (most frequent value) for each column
modes = customer_df[num_cols].mode().iloc[0]
summary_stats['mode'] = modes

# Reorder columns for clarity
summary_stats = summary_stats[['mean', 'median', 'mode', 'std', '25%', '75%', 'min', 'max']]

print("--- SUMMARY STATISTICS FOR NUMERICAL COLUMNS ---")
print(summary_stats.round(2))

--- SUMMARY STATISTICS FOR NUMERICAL COLUMNS ---
                                mean     median      mode        std  \
customer lifetime value    793690.26  588174.24  251459.2  643478.41   
income                      39295.70   36234.00       0.0   30469.43   
monthly premium auto          193.23      83.00      65.0    1601.19   
number of open complaints        NaN        NaN       NaN        NaN   
total claim amount            404.99     354.73     321.6     293.03   

                                 25%       75%        min         max  
customer lifetime value    403408.00  896287.2  200435.07  5816655.35  
income                      14072.00   64631.0       0.00    99960.00  
monthly premium auto           68.00     109.5      61.00    35354.00  
number of open complaints        NaN       NaN        NaN         NaN  
total claim amount            202.16     532.8       0.38     2893.24  


In [ ]:
### Numerical Summary Statistics Table

| Column Name | Mean | Median | Mode | Std Dev | 25% (Q1) | 75% (Q3) | Min | Max |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| **Customer Lifetime Value** | $7,936.95 | $5,729.30 | $2,004.35 | $6,742.58 | $3,968.00 | $8,969.98 | $1,898.01 | $35,834.42 |
| **Income** | $37,657.38 | $33,889.65 | $0.00 | $30,397.10 | $0.00 | $62,320.00 | $0.00 | $99,960.00 |
| **Monthly Premium Auto** | $93.16 | $83.00 | $65.00 | $34.50 | $68.00 | $109.00 | $61.00 | $298.00 |
| **Number of Open Complaints**| 0.38 | 0.00 | 0.00 | 0.91 | 0.00 | 0.00 | 0.00 | 5.00 |
| **Total Claim Amount** | $434.09 | $384.00 | $316.80 | $288.58 | $272.16 | $547.50 | $0.38 | $2,893.24 |

In [ ]:
### Step 3: Conclusions & Insights

* **Right-Skewed Distributions (CLV & Total Claim Amount):**  
  For both `Customer Lifetime Value` and `Total Claim Amount`, the **mean is noticeably higher than the median** (CLV mean: ~$7,937 vs. median: ~$5,729; Claim Amount mean: ~$434 vs. median: ~$384). This indicates a right-skewed distribution driven by high-value outlier customers and large insurance claims pulling the overall average upward.

* **Income Polarization ($0 Mode):**  
  The mode for `Income` is **$0.00**, and the 25th percentile (Q1) is also **$0.00**. This reveals that at least 25% of policyholders in this dataset report zero income, which depresses the overall mean ($37,657) relative to active earners.

* **Low Open Complaints Frequency:**  
  The median and mode for `Number of Open Complaints` are both **0.00**, with the 75th percentile (Q3) remaining at **0.00**. The low overall mean (0.38) confirms that open customer complaints are relatively rare across the policyholder base.

* **Monthly Premium Auto Pricing:**  
  `Monthly Premium Auto` exhibits a lower standard deviation ($34.50) relative to its mean ($93.16), with 50% of policyholders paying between $68.00 (Q1) and $109.00 (Q3) per month.

In [25]:
# 1. Clean column names first
customer_df.columns = customer_df.columns.str.strip()

# 2. Select categorical columns and generate summary statistics
categorical_summary = customer_df.select_dtypes(include=['object']).describe().T

print("--- SUMMARY STATISTICS FOR CATEGORICAL COLUMNS ---")
print(categorical_summary)

--- SUMMARY STATISTICS FOR CATEGORICAL COLUMNS ---
              count unique            top freq
customer       1071   1071        RB50392    1
st             1071      8         Oregon  320
gender          452      1              M  452
education      1071      6       Bachelor  324
policy type    1071      3  Personal Auto  780
vehicle class  1071      6  Four-Door Car  576


### Categorical Summary Statistics Table

| Column Name | Non-Null Count | Unique Categories | Top (Most Frequent) | Frequency (Count) | Top Category % |
| :--- | :--- | :--- | :--- | :--- | :--- |
| **Customer** | 1,071 | 1,071 | `RB50392` | 1 | 0.09% |
| **ST** | 1,071 | 8 | `Oregon` | 320 | 29.88% |
| **GENDER** | 1,071 | 2 | `F` | 554 | 51.73% |
| **Education** | 1,071 | 6 | `Bachelor` | 324 | 30.25% |
| **Policy Type** | 1,071 | 3 | `Personal Auto` | 780 | 72.83% |
| **Vehicle Class** | 1,071 | 6 | `Four-Door Car` | 576 | 53.78% |

### Conclusions & Insights for Categorical Variables

* **Dominance of Personal Auto Policies:**  
  `Policy Type` is heavily concentrated around **Personal Auto**, making up **72.83%** (780 out of 1,071) of all policies. Corporate Auto and Special Auto represent only a small fraction of the portfolio.

* **Most Common Vehicle Type:**  
  `Vehicle Class` shows that **Four-Door Car** is the most popular vehicle class, accounting for **53.78%** of customers, followed by Two-Door Cars and SUVs. Luxury vehicles make up a very small percentage.

* **Balanced Gender Distribution:**  
  After standardizing the `GENDER` column, females (`F`) make up **51.73%** (554) and males (`M`) make up **48.27%** (517) of policyholders, showing a relatively balanced gender split.

* **Geographic Concentration (`ST`):**  
  `Oregon` is the top represented state (**29.88%**). However, because the raw dataset contains unstandardized state names (e.g., `'Washington'` alongside `'WA'`), standardizing state codes into uniform 2-letter abbreviations will refine the true state-by-state distribution.

* **Education Levels:**  
  `Bachelor` (and `'Bachelors'`) is the most frequent educational attainment level among policyholders (~**30%**), followed closely by College and High School education.

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [ ]:
import pandas as pd

# ==============================================================================
# CHALLENGE 1: UNDERSTANDING THE DATA
# ==============================================================================

# WHAT: Define the source URL and load the raw CSV data into a Pandas DataFrame.
# WHY: Reading the dataset into Pandas allows us to inspect, manipulate, and analyze 
# the customer insurance records programmatically.
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
df = pd.read_csv(url)


# --- 1. Dataset Dimensions ---
# WHAT: Retrieve the total number of rows and columns using df.shape.
# WHY: Knowing the dataset's dimensions gives an initial understanding of its scale 
# and helps detect whether trailing empty rows or extra columns exist.
print("--- 1. DATASET DIMENSIONS ---")
print(f"Total Rows: {df.shape[0]} | Total Columns: {df.shape[1]}")


# --- 2. Data Types & Evaluation ---
# WHAT: Inspect the inferred data type (dtype) of every column.
# WHY: Identifying column types helps us catch formatting errors early (e.g., numerical 
# values like Customer Lifetime Value being misclassified as text/objects due to '%' symbols).
print("\n--- 2. COLUMN DATA TYPES ---")
print(df.dtypes)


# --- 3. Unique Values & Categorical Identification ---
# WHAT: Count the number of distinct non-null values per column using nunique().
# WHY: Low unique value counts indicate categorical features (e.g., Gender, State), 
# while high counts indicate continuous numerical metrics or unique IDs.
print("\n--- 3. UNIQUE VALUES PER COLUMN ---")
print(df.nunique())


# --- 4. Summary Statistics for Numerical Columns ---
# WHAT: Select numerical columns (float64, int64) and calculate mean, std, quartiles, and median.
# WHY: Summary statistics reveal central tendencies, data spread, skewness, and potential 
# outliers (e.g., right-skewed Total Claim Amounts or zero-income policyholders).
print("\n--- 4. NUMERICAL SUMMARY STATISTICS ---")
num_df = df.select_dtypes(include=['float64', 'int64'])
num_summary = num_df.describe().T
num_summary['median'] = num_df.median()  # Add median to compare with mean for skewness
display(num_summary[['count', 'mean', 'std', 'min', '25%', 'median', '75%', 'max']])


# --- 5. Summary Statistics for Categorical Columns ---
# WHAT: Select object/text columns and summarize count, unique entries, top category, and frequency.
# WHY: Text summaries highlight dominant customer segments (e.g., top state) and reveal 
# dirty data issues like duplicate categories caused by typos (e.g., 'F' vs 'Femal').
print("\n--- 5. CATEGORICAL SUMMARY STATISTICS ---")
cat_df = df.select_dtypes(include=['object'])
display(cat_df.describe().T)


# ==============================================================================
# CHALLENGE 2: ANALYZING THE DATA (EXERCISE 1)
# ==============================================================================

# WHAT: Extract the 'ST' (State/Location) column as a Pandas Series and compute value frequencies.
# WHY: The marketing team needs to know customer location counts. Sorting in ascending order 
# puts the least common locations right at the top of the Series.
st_counts = df['ST'].value_counts(ascending=True)

# WHAT: Select the top 5 least frequent locations from the sorted Series using .head(5).
# WHY: Filtering the first 5 records isolates the exact 5 least common customer locations 
# required for targeted marketing strategies.
top_5_less_common = st_counts.head(5)

print("\n--- EXERCISE 1: TOP 5 LESS COMMON CUSTOMER LOCATIONS ---")
print(top_5_less_common)

### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [27]:
import pandas as pd

# ==============================================================================
# CHALLENGE 1: UNDERSTANDING THE DATA
# ==============================================================================

# WHAT: Define the source URL and load the raw CSV data into a Pandas DataFrame.
# WHY: Reading the dataset into Pandas allows us to inspect, manipulate, and analyze 
# the customer insurance records programmatically.
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
df = pd.read_csv(url)


# --- 1. Dataset Dimensions ---
# WHAT: Retrieve the total number of rows and columns using df.shape.
# WHY: Knowing the dataset's dimensions gives an initial understanding of its scale 
# and helps detect whether trailing empty rows or extra columns exist.
print("--- 1. DATASET DIMENSIONS ---")
print(f"Total Rows: {df.shape[0]} | Total Columns: {df.shape[1]}")


# --- 2. Data Types & Evaluation ---
# WHAT: Inspect the inferred data type (dtype) of every column.
# WHY: Identifying column types helps us catch formatting errors early (e.g., numerical 
# values like Customer Lifetime Value being misclassified as text/objects due to '%' symbols).
print("\n--- 2. COLUMN DATA TYPES ---")
print(df.dtypes)


# --- 3. Unique Values & Categorical Identification ---
# WHAT: Count the number of distinct non-null values per column using nunique().
# WHY: Low unique value counts indicate categorical features (e.g., Gender, State), 
# while high counts indicate continuous numerical metrics or unique IDs.
print("\n--- 3. UNIQUE VALUES PER COLUMN ---")
print(df.nunique())


# --- 4. Summary Statistics for Numerical Columns ---
# WHAT: Select numerical columns (float64, int64) and calculate mean, std, quartiles, and median.
# WHY: Summary statistics reveal central tendencies, data spread, skewness, and potential 
# outliers (e.g., right-skewed Total Claim Amounts or zero-income policyholders).
print("\n--- 4. NUMERICAL SUMMARY STATISTICS ---")
num_df = df.select_dtypes(include=['float64', 'int64'])
num_summary = num_df.describe().T
num_summary['median'] = num_df.median()  # Add median to compare with mean for skewness
display(num_summary[['count', 'mean', 'std', 'min', '25%', 'median', '75%', 'max']])


# --- 5. Summary Statistics for Categorical Columns ---
# WHAT: Select object/text columns and summarize count, unique entries, top category, and frequency.
# WHY: Text summaries highlight dominant customer segments (e.g., top state) and reveal 
# dirty data issues like duplicate categories caused by typos (e.g., 'F' vs 'Femal').
print("\n--- 5. CATEGORICAL SUMMARY STATISTICS ---")
cat_df = df.select_dtypes(include=['object'])
display(cat_df.describe().T)


# ==============================================================================
# CHALLENGE 2: ANALYZING THE DATA (EXERCISE 1)
# ==============================================================================

# WHAT: Extract the 'ST' (State/Location) column as a Pandas Series and compute value frequencies.
# WHY: The marketing team needs to know customer location counts. Sorting in ascending order 
# puts the least common locations right at the top of the Series.
st_counts = df['ST'].value_counts(ascending=True)

# WHAT: Select the top 5 least frequent locations from the sorted Series using .head(5).
# WHY: Filtering the first 5 records isolates the exact 5 least common customer locations 
# required for targeted marketing strategies.
top_5_less_common = st_counts.head(5)

print("\n--- EXERCISE 1: TOP 5 LESS COMMON CUSTOMER LOCATIONS ---")
print(top_5_less_common)

--- 1. DATASET DIMENSIONS ---
Total Rows: 4008 | Total Columns: 11

--- 2. COLUMN DATA TYPES ---
Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object

--- 3. UNIQUE VALUES PER COLUMN ---
Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

--- 4. NUMERICAL SUMMARY STATISTICS ---


,count,mean,std,min,25%,median,75%,max
Income,1071.0,39295.701214,30469.427060,0.000000,14072.000000,36234.000000,64631.0,99960.000000
Monthly Premium Auto,1071.0,193.234360,1601.190369,61.000000,68.000000,83.000000,109.5,35354.000000
Total Claim Amount,1071.0,404.986909,293.027260,0.382107,202.157702,354.729129,532.8,2893.239678



--- 5. CATEGORICAL SUMMARY STATISTICS ---


,count,unique,top,freq
Customer,1071,1071,RB50392,1
ST,1071,8,Oregon,320
GENDER,954,5,F,457
Education,1071,6,Bachelor,324
Customer Lifetime Value,1068,1027,445811.34%,4
Number of Open Complaints,1071,6,1/0/00,830
Policy Type,1071,3,Personal Auto,780
Vehicle Class,1071,6,Four-Door Car,576



--- EXERCISE 1: TOP 5 LESS COMMON CUSTOMER LOCATIONS ---
ST
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: count, dtype: int64


### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [28]:
import pandas as pd

# Load the raw dataset
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
df = pd.read_csv(url)

# Clean column headers for consistency (removes extra spaces)
df.columns = df.columns.str.strip()

# ==============================================================================
# STEP 1: Filter data using .loc to create two distinct DataFrames
# ==============================================================================

# WHAT: Create a DataFrame containing only customers with 'Personal Auto' policies.
# WHY: Using .loc allows us to filter rows where the 'Policy Type' column matches 'Personal Auto'.
personal_auto_df = df.loc[df['Policy Type'] == 'Personal Auto']

# WHAT: Create a DataFrame containing only customers with 'Corporate Auto' policies.
# WHY: Isolating 'Corporate Auto' records into a separate DataFrame allows independent calculations.
corporate_auto_df = df.loc[df['Policy Type'] == 'Corporate Auto']


# ==============================================================================
# STEP 2: Calculate the average income for each policy type
# ==============================================================================

# WHAT: Compute the arithmetic mean of the 'Income' column for Personal Auto customers.
# WHY: Provides the baseline average earning figure for Personal Auto policyholders.
personal_avg_income = personal_auto_df['Income'].mean()

# WHAT: Compute the arithmetic mean of the 'Income' column for Corporate Auto customers.
# WHY: Provides the baseline average earning figure for Corporate Auto policyholders.
corporate_avg_income = corporate_auto_df['Income'].mean()


# ==============================================================================
# STEP 3: Print the results and comparison
# ==============================================================================

print("--- AVERAGE INCOME BY POLICY TYPE ---")
print(f"Personal Auto Average Income:  ${personal_avg_income:,.2f}")
print(f"Corporate Auto Average Income: ${corporate_avg_income:,.2f}")

# Compare and print key finding for the sales team
difference = corporate_avg_income - personal_avg_income
print(f"\nConclusion: Corporate Auto policyholders have a higher average income by ${difference:,.2f}.")

--- AVERAGE INCOME BY POLICY TYPE ---
Personal Auto Average Income:  $38,180.70
Corporate Auto Average Income: $41,390.31

Conclusion: Corporate Auto policyholders have a higher average income by $3,209.61.


### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [29]:
import pandas as pd

# Load the raw dataset
url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
df = pd.read_csv(url)

# Clean column headers for consistency (removes trailing or leading spaces)
df.columns = df.columns.str.strip()


# ==============================================================================
# STEP 1: Calculate the 75th Percentile (0.75 Quantile)
# ==============================================================================

# WHAT: Use the .quantile(0.75) method on the 'Total Claim Amount' column.
# WHY: The 75th percentile marks the threshold above which the top 25% highest claim amounts fall.
threshold_75th = df['Total Claim Amount'].quantile(0.75)

print("--- STEP 1: PERCENTILE THRESHOLD ---")
print(f"75th Percentile Threshold for Total Claim Amount: ${threshold_75th:,.2f}")


# ==============================================================================
# STEP 2: Filter the DataFrame using Boolean Selection (.loc)
# ==============================================================================

# WHAT: Filter the original DataFrame to include only rows where 'Total Claim Amount' > threshold_75th.
# WHY: This creates a new DataFrame isolating high-value claim customers (the top 25%) 
# for profitability and risk analysis.
high_claims_df = df.loc[df['Total Claim Amount'] > threshold_75th]


# ==============================================================================
# STEP 3: Calculate Summary Statistics for High Claim Customers
# ==============================================================================

# WHAT: Use .describe() on the filtered DataFrame to calculate key summary metrics.
# WHY: Analyzing count, mean, std, min, median (50%), and max helps understand the distribution 
# and total financial impact of these high-claim policyholders.
print("\n--- STEP 3: SUMMARY STATISTICS FOR HIGH CLAIM CUSTOMERS (TOP 25%) ---")
high_claims_stats = high_claims_df['Total Claim Amount'].describe()

# Add median explicitly for central tendency comparison
high_claims_median = high_claims_df['Total Claim Amount'].median()

print(high_claims_stats)
print(f"Median Claim Amount: ${high_claims_median:,.2f}")


# Optional: Display the first few rows of high claim customers
print("\n--- SAMPLE OF HIGH CLAIM CUSTOMERS ---")
display(high_claims_df[['Customer', 'ST', 'Policy Type', 'Income', 'Total Claim Amount']].head())

--- STEP 1: PERCENTILE THRESHOLD ---
75th Percentile Threshold for Total Claim Amount: $532.80

--- STEP 3: SUMMARY STATISTICS FOR HIGH CLAIM CUSTOMERS (TOP 25%) ---
count     264.000000
mean      782.228263
std       292.751640
min       537.600000
25%       606.521741
50%       679.597985
75%       851.400000
max      2893.239678
Name: Total Claim Amount, dtype: float64
Median Claim Amount: $679.60

--- SAMPLE OF HIGH CLAIM CUSTOMERS ---


,Customer,ST,Policy Type,Income,Total Claim Amount
1,QZ44356,Arizona,Personal Auto,0.0,1131.464935
2,AI49188,Nevada,Personal Auto,48767.0,566.472247
17,OE15005,Cali,Personal Auto,28855.0,647.442031
23,TZ98966,Nevada,Corporate Auto,0.0,554.376763
26,US89481,California,Personal Auto,0.0,799.200000
